Bucketeer2/Sources/Internal/Remote/ApiClientImpl.swift

```diff
@@ -0,0 +1,128 @@
+            }
+        )
+    }
+
+    func registerEvents(events: [Event], completion: ((Result<RegisterEventsResponse, BKTError>) -> Void)?) {
+        let requestBody = RegisterEventsRequestBody(
+            events: events
+        )
+        logger?.debug(message: "[API] Register events: \(requestBody)")
+        send(
+            requestBody: requestBody,
+            path: "v1/gateway/events",
+            completion: { (result: Result<(RegisterEventsResponse, URLResponse), Error>) in
+                switch result {
+                case .success((let response, _)):
+                    completion?(.success(response))
+                case .failure(let error):
+                    completion?(.failure(.init(error: error)))
+                }
+            }
+        )
+    }
+
+    func send<RequestBody: Encodable, Response: Decodable>(requestBody: RequestBody, path: String, timeoutMillis: Int64? = nil, completion: ((Result<(Response, URLResponse), Error>) -> Void)?) {
+
+        do {
+            let body = try JSONEncoder().encode(requestBody)
+            var request = URLRequest(url: endpoint.appendingPathComponent(path))
+            request.httpMethod = "POST"
+            request.allHTTPHeaderFields = [
+                "Authorization": self.apiKey
+            ]
+            request.httpBody = body
+            request.timeoutInterval = TimeInterval(timeoutMillis ?? defaultRequestTimeoutMills) / 1000
+
+            session.task(with: request) { data, urlResponse, error in
+                guard let data = data else {
+                    guard let error = error else {
+                        completion?(.failure(ResponseError.unknown(urlResponse)))
+                        return
+                    }
+                    completion?(.failure(error))
+                    return
+                }
+
+                guard let urlResponse = urlResponse as? HTTPURLResponse else {
+                    completion?(.failure(ResponseError.unknown(urlResponse)))
+                    return
+                }
+                do {
+                    guard 200..<300 ~= urlResponse.statusCode else {
+                        let response: ErrorResponse? = try JSONDecoder().decode(ErrorResponse.self, from: data)
+                        let error = ResponseError.unacceptableCode(code: urlResponse.statusCode, response: response)
+                        completion?(.failure(error))
+                        return
+                    }
+                    let response = try JSONDecoder().decode(Response.self, from: data)
+                    completion?(.success((response, urlResponse)))
+                } catch let error {
+                    completion?(.failure(error))
+                }
+            }
+        } catch let error {
+            completion?(.failure(error))
+        }
+    }
+}
+
+enum ResponseError: Error {
+    case unknown(URLResponse?)
+    case unacceptableCode(code: Int, response: ErrorResponse?)
```

Comment:
I think we could use another name for this.
Perhaps, `checkErrorStatusCode` ? or `getErrorByStatusCode` ?
What do you think?

The rest looks good!

Reply:
Oh ok, it's common to use it in the swift development.

Thanks!